File: dotnet-impl/net/terria/registry/terraform/RegistryService.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;

namespace net.terria.registry.terraform
{
    /// <summary>
    /// Service that scans local filesystem and builds in-memory indexes for provider and module metadata.
    /// </summary>
    public class RegistryService
    {
        private readonly ILogger<RegistryService> _logger;
        private readonly Dictionary<string, ProviderIndex> _providers = new();
        private readonly Dictionary<string, ModuleIndex> _modules = new();
        private readonly string _storagePath = string.Empty;

        public RegistryService(ILogger<RegistryService> logger, IConfiguration config)
        {
            _logger = logger;
            _storagePath = config.GetValue<string>("Registry:StoragePath") ?? "../registry-storage";
            Init();
        }

        /// <summary>
        /// Initialize in-memory service indexes from local registry storage on disk.
        /// </summary>
        private void Init()
        {
            var root = Path.Combine(_storagePath);
            if (!Directory.Exists(root))
            {
                _logger.LogWarning("Registry storage path does not exist: {Path}", root);
                return;
            }

            ScanProviders(Path.Combine(root, "providers"));
            ScanModules(Path.Combine(root, "modules"));

            _logger.LogInformation("Registry indexes rebuilt. Providers: {ProviderCount}, Modules: {ModuleCount}", _providers.Count, _modules.Count);
        }

        /// <summary>
        /// Scan the top-level providers path (e.g. <c>registry-storage/providers</c>) and index each registry subtree.
        /// </summary>
        /// <param name="providersPath">Filesystem path for provider registry roots.</param>
        private void ScanProviders(string providersPath)
        {
            if (!Directory.Exists(providersPath)) return;

            foreach (var registryDir in Directory.EnumerateDirectories(providersPath))
            {
                var registry = Path.GetFileName(registryDir);
                ScanRegistryProviders(registryDir, registry);
            }
        }

        /// <summary>
        /// Scan a provider registry root path (e.g. <c>registry.terraform.io</c>) and process namespaces.
        /// </summary>
        /// <param name="registryPath">Registry directory on disk</param>
        /// <param name="registry">Registry ID</param>
        private void ScanRegistryProviders(string registryPath, string registry)
        {
            foreach (var namespaceDir in Directory.EnumerateDirectories(registryPath))
            {
                var namespaceName = Path.GetFileName(namespaceDir);
                if (string.IsNullOrEmpty(namespaceName) || namespaceName.Equals("namespaces.json", StringComparison.OrdinalIgnoreCase))
                {
                    continue;
                }

                ScanNamespaceProviders(namespaceDir, registry, namespaceName);
            }
        }

        /// <summary>
        /// Scan a provider namespace path and index provider packages.
        /// </summary>
        /// <param name="namespacePath">Filesystem path for this namespace</param>
        /// <param name="registry">Registry ID</param>
        /// <param name="namespaceName">Namespace name</param>
        private void ScanNamespaceProviders(string namespacePath, string registry, string namespaceName)
        {
            foreach (var providerDir in Directory.EnumerateDirectories(namespacePath))
            {
                var providerName = Path.GetFileName(providerDir);
                var index = ScanProvider(providerDir, registry, namespaceName, providerName);
                _providers[$"{registry}/{namespaceName}/{providerName}"] = index;
            }
        }

        /// <summary>
        /// Scan one provider directory and collect available version entries with platform architecture metadata.
        /// </summary>
        /// <param name="providerPath">Path to provider directory</param>
        /// <param name="registry">Registry id</param>
        /// <param name="namespaceName">Namespace id</param>
        /// <param name="name">Provider name</param>
        /// <returns>Provider index object</returns>
        private ProviderIndex ScanProvider(string providerPath, string registry, string namespaceName, string name)
        {
            // Accept any version directory under provider (including symlink / alias names) as-is.
            var versions = new Dictionary<string, List<Platform>>();

            foreach (var versionDir in Directory.EnumerateDirectories(providerPath))
            {
                var versionStr = Path.GetFileName(versionDir);
                var platforms = ScanPlatforms(versionDir);
                versions[versionStr] = platforms;
                _logger.LogDebug("Discovered provider {Registry}/{Namespace}/{Name}/{Version} with {PlatformCount} platforms", registry, namespaceName, name, versionStr, platforms.Count);
            }

            return new ProviderIndex(registry, namespaceName, name, versions);
        }

        /// <summary>
        /// Scan OS/arch directories under a specific version and produce Platform entries.
        /// </summary>
        /// <param name="versionPath">Filesystem version directory path</param>
        /// <returns>List of Platform metadata</returns>
        private List<Platform> ScanPlatforms(string versionPath)
        {
            var platforms = new List<Platform>();

            if (!Directory.Exists(versionPath)) return platforms;

            foreach (var osDir in Directory.EnumerateDirectories(versionPath))
            {
                var os = Path.GetFileName(osDir);
                foreach (var archDir in Directory.EnumerateDirectories(osDir))
                {
                    var arch = Path.GetFileName(archDir);
                    platforms.Add(new Platform(os, arch));
                }
            }

            return platforms;
        }

        /// <summary>
        /// Scan the top-level modules path (e.g. <c>registry-storage/modules</c>) and index each registry subtree.
        /// </summary>
        /// <param name="modulesPath">Filesystem path for module registry roots</param>
        private void ScanModules(string modulesPath)
        {
            if (!Directory.Exists(modulesPath)) return;

            foreach (var registryDir in Directory.EnumerateDirectories(modulesPath))
            {
                var registry = Path.GetFileName(registryDir);
                ScanRegistryModules(registryDir, registry);
            }
        }

        private void ScanRegistryModules(string registryPath, string registry)
        {
            foreach (var namespaceDir in Directory.EnumerateDirectories(registryPath))
            {
                var namespaceName = Path.GetFileName(namespaceDir);
                if (string.IsNullOrEmpty(namespaceName) || namespaceName.Equals("namespaces.json", StringComparison.OrdinalIgnoreCase))
                {
                    continue;
                }

                ScanNamespaceModules(namespaceDir, registry, namespaceName);
            }
        }

        private void ScanNamespaceModules(string namespacePath, string registry, string namespaceName)
        {
            foreach (var moduleDir in Directory.EnumerateDirectories(namespacePath))
            {
                var moduleName = Path.GetFileName(moduleDir);
                var moduleIndex = ScanModule(moduleDir, registry, namespaceName, moduleName);
                _modules[$"{registry}/{namespaceName}/{moduleName}"] = moduleIndex;
            }
        }

        private ModuleIndex ScanModule(string modulePath, string registry, string namespaceName, string name)
        {
            // Keep all module version directories as provided by filesystem.
            var versions = new Dictionary<string, ModuleVersion>();

            foreach (var versionDir in Directory.EnumerateDirectories(modulePath))
            {
                var versionStr = Path.GetFileName(versionDir);
                versions[versionStr] = new ModuleVersion(versionStr, ScanPlatforms(versionDir));
                _logger.LogDebug("Discovered module {Registry}/{Namespace}/{Name}/{Version} with {PlatformCount} platforms", registry, namespaceName, name, versionStr, versions[versionStr].Platforms.Count);
            }

            return new ModuleIndex(registry, namespaceName, name, versions);
        }

        /// <summary>
        /// Returns provider index for registry/namespace/name. Null if not found.
        /// </summary>
        public ProviderIndex? GetProvider(string registry, string namespaceName, string name)
        {
            _providers.TryGetValue($"{registry}/{namespaceName}/{name}", out var index);
            return index;
        }

        /// <summary>
        /// Returns module index for registry/namespace/name. Null if not found.
        /// </summary>
        public ModuleIndex? GetModule(string registry, string namespaceName, string name)
        {
            _modules.TryGetValue($"{registry}/{namespaceName}/{name}", out var index);
            return index;
        }

        /// <summary>
        /// Get all providers (for testing).
        /// </summary>
        public IReadOnlyDictionary<string, ProviderIndex> GetProviders() => _providers;

        /// <summary>
        /// Get all modules (for testing).
        /// </summary>
        public IReadOnlyDictionary<string, ModuleIndex> GetModules() => _modules;
    }

    /// <summary>
    /// Represents a platform (OS/architecture) where a provider or module is available.
    /// </summary>
    public record Platform(string Os, string Arch);

    /// <summary>
    /// Represents a specific version of a Terraform module with its available platforms.
    /// </summary>
    public record ModuleVersion(string Version, List<Platform> Platforms);

    /// <summary>
    /// Index of all available versions and platforms for a provider.
    /// </summary>
    public record ProviderIndex(string Registry, string Namespace, string Name, Dictionary<string, List<Platform>> Versions);

    /// <summary>
    /// Index of all available versions and platforms for a module.
    /// </summary>
    public record ModuleIndex(string Registry, string Namespace, string Name, Dictionary<string, ModuleVersion> Versions);
}

File: java-impl/src/main/java/net/terria/registry/terraform/RegistryService.java
package net.terria.registry.terraform;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import jakarta.annotation.PostConstruct;

@Service
public class RegistryService {

    private static final Logger logger = LoggerFactory.getLogger(RegistryService.class);

    /**
     * Root path where on-disk registry storage is mounted.
     * Can be configured through application properties.
     */
    @Value("${registry.storage.path:../registry}")
    private String storagePath;

    /**
     * In-memory index keyed by "registry/namespaceName/name" for providers.
     */
    private final Map<String, ProviderIndex> providers = new ConcurrentHashMap<>();

    /**
     * In-memory index keyed by "registry/namespaceName/name" for modules.
     */
    private final Map<String, ModuleIndex> modules = new ConcurrentHashMap<>();

    /**
     * Initialize the in-memory index by scanning filesystem at startup.
     * This mirrors dynamic .json generation from repo content.
     */
    @PostConstruct
    public void init() throws IOException {
        Path root = Paths.get(storagePath);
        if (!Files.exists(root)) {
            logger.warn("Registry storage path does not exist: {}", root);
            return;
        }

        // Scan the terraform registry structure
        Path terraformRegistry = root.resolve("registry.terraform.io");
        if (Files.exists(terraformRegistry)) {
            scanTerraformRegistry(terraformRegistry);
        }

        logger.info("Registry indexes rebuilt. Providers: {}, Modules: {}", providers.size(), modules.size());
    }

    /**
     * Scan a specific terraform registry directory structure.
     */
    private void scanTerraformRegistry(Path registryPath) throws IOException {
        String registry = "registry.terraform.io";

        // Scan providers under registry.terraform.io/providers/
        Path providersPath = registryPath.resolve("providers");
        if (Files.exists(providersPath)) {
            scanRegistryProviders(providersPath, registry);
        }

        // Scan modules under registry.terraform.io/modules/
        Path modulesPath = registryPath.resolve("modules");
        if (Files.exists(modulesPath)) {
            scanRegistryModules(modulesPath, registry);
        }
    }

    /**
     * Scan provider storage directories and build in-memory provider index.
     */
    private void scanProviders(Path providersPath) throws IOException {
        if (!Files.exists(providersPath)) return;

        // Top-level under providers: registry namespaces like registry.terraform.io
        Files.walk(providersPath, 1)
            .filter(Files::isDirectory)
            .filter(p -> !p.equals(providersPath))
            .forEach(registryPath -> {
                String registry = registryPath.getFileName().toString();
                try {
                    scanRegistryProviders(registryPath, registry);
                } catch (IOException e) {
                    logger.error("Error scanning providers in registry {}", registry, e);
                }
            });
    }

    private void scanRegistryProviders(Path registryPath, String registry) throws IOException {
        Path namespacesPath = registryPath.resolve("namespaces.json");
        if (!Files.exists(namespacesPath)) return; // Skip if no namespaces.json

        Files.list(registryPath)
            .filter(Files::isDirectory)
            .filter(p -> !p.getFileName().toString().equals("namespaces.json"))
            .forEach(namespacePath -> {
                String namespaceName = namespacePath.getFileName().toString();
                try {
                    scanNamespaceProviders(namespacePath, registry, namespaceName);
                } catch (IOException e) {
                    logger.error("Error scanning namespace {} in registry {}", namespaceName, registry, e);
                }
            });
    }

    private void scanNamespaceProviders(Path namespacePath, String registry, String namespaceName) throws IOException {
        Files.list(namespacePath)
            .filter(Files::isDirectory)
            .forEach(providerPath -> {
                String providerName = providerPath.getFileName().toString();
                try {
                    ProviderIndex index = scanProvider(providerPath, registry, namespaceName, providerName);
                    String key = registry + "/" + namespaceName + "/" + providerName;
                    providers.put(key, index);
                } catch (Exception e) {
                    logger.error("Error scanning provider {}/{}/{}", registry, namespaceName, providerName, e);
                }
            });
    }

    private ProviderIndex scanProvider(Path providerPath, String registry, String namespaceName, String name) throws IOException {
        // Keep raw folder names to reflect exactly what exists on disk.
        // Accept symlink/alias version directories as-is by name.
        Map<String, List<Platform>> versions = new LinkedHashMap<>();

        Files.list(providerPath)
            .filter(Files::isDirectory)
            .forEach(versionPath -> {
                String versionStr = versionPath.getFileName().toString();
                try {
                    List<Platform> platforms = scanPlatforms(versionPath);
                    versions.put(versionStr, platforms);
                    logger.debug("Discovered provider {}/{}/{}/{} with {} platform entries", registry, namespaceName, name, versionStr, platforms.size());
                } catch (IOException ex) {
                    logger.warn("Failed to scan provider version {}: {}", versionStr, ex.getMessage());
                }
            });

        return new ProviderIndex(registry, namespaceName, name, versions);
    }

    private List<Platform> scanPlatforms(Path versionPath) throws IOException {
        if (!Files.exists(versionPath)) return Collections.emptyList();

        // Magic platform directory detection in each provider version.
        return Files.list(versionPath)
            .filter(Files::isDirectory)
            .flatMap(osPath -> {
                String os = osPath.getFileName().toString();
                try {
                    return Files.list(osPath)
                        .filter(Files::isDirectory)
                        .map(archPath -> new Platform(os, archPath.getFileName().toString()));
                } catch (IOException ex) {
                    logger.warn("Failed to scan arch directories for os {}: {}", os, ex.getMessage());
                    return java.util.stream.Stream.empty();
                }
            })
            .collect(Collectors.toList());
    }

    /**
     * Scan module storage directories and build in-memory module index.
     */
    private void scanModules(Path modulesPath) throws IOException {
        if (!Files.exists(modulesPath)) return;

        Files.walk(modulesPath, 1)
            .filter(Files::isDirectory)
            .filter(p -> !p.equals(modulesPath))
            .forEach(registryPath -> {
                String registry = registryPath.getFileName().toString();
                try {
                    scanRegistryModules(registryPath, registry);
                } catch (IOException e) {
                    logger.error("Error scanning modules in registry {}", registry, e);
                }
            });
    }

    private void scanRegistryModules(Path registryPath, String registry) throws IOException {
        Path namespacesPath = registryPath.resolve("namespaces.json");
        if (!Files.exists(namespacesPath)) return;

        Files.list(registryPath)
            .filter(Files::isDirectory)
            .filter(p -> !p.getFileName().toString().equals("namespaces.json"))
            .forEach(namespacePath -> {
                String namespaceName = namespacePath.getFileName().toString();
                try {
                    scanNamespaceModules(namespacePath, registry, namespaceName);
                } catch (IOException e) {
                    logger.error("Error scanning module namespace {} in registry {}", namespaceName, registry, e);
                }
            });
    }

    private void scanNamespaceModules(Path namespacePath, String registry, String namespaceName) throws IOException {
        Files.list(namespacePath)
            .filter(Files::isDirectory)
            .forEach(modulePath -> {
                String moduleName = modulePath.getFileName().toString();
                try {
                    ModuleIndex moduleIndex = scanModule(modulePath, registry, namespaceName, moduleName);
                    String key = registry + "/" + namespaceName + "/" + moduleName;
                    modules.put(key, moduleIndex);
                } catch (Exception e) {
                    logger.error("Error scanning module {}/{}/{}", registry, namespaceName, moduleName, e);
                }
            });
    }

    private ModuleIndex scanModule(Path modulePath, String registry, String namespaceName, String name) throws IOException {
        // Keep raw module versions as directory names to mirror filesystem; no strict semver required.
        Map<String, ModuleVersion> versions = new LinkedHashMap<>();

        Files.list(modulePath)
            .filter(Files::isDirectory)
            .forEach(versionPath -> {
                String versionStr = versionPath.getFileName().toString();
                try {
                    ModuleVersion moduleVersion = scanModuleVersion(versionPath, versionStr);
                    versions.put(versionStr, moduleVersion);
                    logger.debug("Discovered module {}/{}/{}/{} with {} system entries", registry, namespaceName, name, versionStr, moduleVersion.systems.size());
                } catch (IOException ex) {
                    logger.warn("Failed to scan module version {}: {}", versionStr, ex.getMessage());
                }
            });

        return new ModuleIndex(registry, namespaceName, name, versions);
    }

    private ModuleVersion scanModuleVersion(Path versionPath, String version) throws IOException {
        List<String> systems = scanModuleSystems(versionPath);
        return new ModuleVersion(version, systems);
    }

    private List<String> scanModuleSystems(Path versionPath) throws IOException {
        if (!Files.isDirectory(versionPath)) {
            return List.of();
        }

        try (var stream = Files.list(versionPath)) {
            return stream
                .filter(Files::isDirectory)
                .map(path -> path.getFileName().toString())
                .toList();
        }
    }

    /**
     * Lookup provider index by full provider path key.
     */
    public Optional<ProviderIndex> getProvider(String registry, String namespaceName, String name) {
        return Optional.ofNullable(providers.get(registry + "/" + namespaceName + "/" + name));
    }

    /**
     * Lookup module index by full module path key.
     */
    public Optional<ModuleIndex> getModule(String registry, String namespaceName, String name) {
        return Optional.ofNullable(modules.get(registry + "/" + namespaceName + "/" + name));
    }

    /**
     * Get all providers (for testing).
     */
    public Map<String, ProviderIndex> getProviders() {
        return providers;
    }

    /**
     * Get all modules (for testing).
     */
    public Map<String, ModuleIndex> getModules() {
        return modules;
    }

    /**
     * Provider index data model.
     */
    public static class ProviderIndex {
        public final String registry;
        public final String namespaceName;
        public final String name;
        public final Map<String, List<Platform>> versions;

        public ProviderIndex(String registry, String namespaceName, String name, Map<String, List<Platform>> versions) {
            this.registry = registry;
            this.namespaceName = namespaceName;
            this.name = name;
            this.versions = versions;
        }
    }

    public static class ModuleIndex {
        public final String registry;
        public final String namespaceName;
        public final String name;
        public final Map<String, ModuleVersion> versions;

        public ModuleIndex(String registry, String namespaceName, String name, Map<String, ModuleVersion> versions) {
            this.registry = registry;
            this.namespaceName = namespaceName;
            this.name = name;
            this.versions = versions;
        }
    }

    public static class ModuleVersion {
        public final String version;
        public final List<String> systems;

        public ModuleVersion(String version, List<String> systems) {
            this.version = version;
            this.systems = systems;
        }
    }

    public static class Platform {
        public final String os;
        public final String arch;

        public Platform(String os, String arch) {
            this.os = os;
            this.arch = arch;
        }
    }
}